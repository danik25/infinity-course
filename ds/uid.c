#include "uid.h"

size_t cnt = 0;
UID_t g_bad_uid = {0, {0}, 0};
/*
UID_t g_bad_uid = UIDCreate();
g_bad_uid.counter += 1;
*/
UID_t UIDCreate()
{
    UID_t new_uid = {0, {0}, 0};
    new_uid.pid = getpid();
    gettimeofday(&new_uid.time, NULL);
    new_uid.counter = ++cnt;

    return (new_uid);
}

int UIDIsSame(UID_t uid1, UID_t uid2)
{
    return(uid1.pid == uid2.pid && uid1.time.tv_sec == uid2.time.tv_sec
                    && uid1.time.tv_usec == uid2.time.tv_usec 
                    && uid1.counter == uid2.counter);
}

int UIDIsBad(UID_t uid)
{
    
    return(UIDIsSame(g_bad_uid, uid));
}