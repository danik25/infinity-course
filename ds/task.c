/****************************************************************
*                   Scheduler                                   *
* Related files: scheduler.c, scheduler_test.c scheduler.h      *
*                task.c, task.h uid.h, uid.c                    *
* #Version: V 1.3                                               *
* Writer: Dani Benjamin                                         *
* Reviewer: Moshe Greenberg                                     *
* Last update: 21/04/19                                         *
****************************************************************/
#include <time.h> /* time_t*/
#include <stddef.h> /* size_t */
#include <stdlib.h> /*malloc*/
#include <assert.h> /*assert*/

#include "uid.h"   /* UID_t */
#include "task.h" /* task_t */


struct task {
    
    UID_t uid;    
    task_function_t *sched_user_func;
    void *param;
    time_t act_time;
	size_t interval; /* seconds */
};

task_t *TaskCreate(task_function_t act_func, void *param, size_t interval)
{
    task_t *task = (task_t *) malloc(sizeof(task_t));
    assert(act_func);
    if(NULL == task)
    {
        return (NULL);
    }
    task->uid = UIDCreate();
    task->sched_user_func = act_func;
    task->param = param;
    task->interval = interval;
    task->act_time = 0;
    return (task);
}

void TaskDestroy(task_t *task)
{
    assert(task);
    free(task);
    task = NULL;
}

int TaskRun(task_t *task)
{
    assert(task);
    return((*task->sched_user_func)(task->param));
}

void TaskTimeUpdate(task_t *task)
{
    assert(task);
    task->act_time = time(NULL) + task->interval;
}

time_t TaskGetTimeToRun(task_t *task)
{
    assert(task);
    return(task->act_time);
}

UID_t TaskGetUID(task_t *task)
{
    assert(task);
    return(task->uid);
}
