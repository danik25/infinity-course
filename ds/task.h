#ifndef _TASK_H_
#define _TASK_H_

/****************************************************************
*                   Scheduler                                   *
* Related files: scheduler.c, scheduler_test.c scheduler.h      *
*                task.c, task.h uid.h, uid.c                    *
* #Version: V 1.3                                               *
* Writer: Dani Benjamin                                         *
* Reviewer: Moshe Greenberg                                     *
* Last update: 21/04/19                                         *
****************************************************************/
#include <time.h> /* time_t*/
#include <stddef.h> /* size_t */

#include "uid.h" /* UID_t */
             

typedef struct task task_t;
typedef int (task_function_t)(void *param);


/* Description: creates a task.
 *
 * Params: 
 *         param - the parameter of the act_func decided by user.
 *         interval - task schduled frequency.
 *         act_func - user function to be executed. 
 *                                      returns: 0 - success (and run again)
 *                                               1 - end (don't run again)
 *                                              -1 - fail (and don't run again)
 * Return Value: pointer to the new task.
 *         
*/
task_t *TaskCreate(task_function_t act_func, void *param, size_t interval);
                                                        



/* Description: destroys the task. frees all allocated memory.
 *
 * Params: task - task to destroy.
 *        
 *         
*/
void TaskDestroy(task_t *task);


/* Description: execute a task.
 *
 * Params: task - task to execute.
 *        
 * Return Value: status of the function that was activated.
 *         
*/
int TaskRun(task_t *task);


/* Description: updates the time stamp of the task.
 *
 * Params: task - the task to update.
 *        
 *         
*/
void TaskTimeUpdate(task_t *task);


/* Description: gets the time stamp of the task.
 *
 * Params: task - the task to take the time from.
 *        
 * Return Value: the time stamp of the task.
 *         
*/
time_t TaskGetTimeToRun(task_t *task);



/* Description: gets the UID of the task.
 *
 * Params: task - the task to take the time from.
 *        
 * Return Value: the UID of the task.
 *         
*/
UID_t TaskGetUID(task_t *task);

#endif /* _TASK_H_ */