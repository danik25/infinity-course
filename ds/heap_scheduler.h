#ifndef  _SCHEDULER_H_
#define  _SCHEDULER_H_


/****************************************************************
*                   Scheduler                                   *
* Related files: scheduler.c, scheduler_test.c scheduler.h      *
*                task.c, task.h uid.h, uid.c                    *
* #Version: V 1.3                                               *
* Writer: Dani Benjamin                                         *
* Reviewer: Moshe Greenberg                                     *
* Last update: 21/04/19                                         *
****************************************************************/
/* API DESCRIPTION
 *
 * Scheduler is a program that schedules tasks and executes them at a time set
 * by the user.
 *
 *
*/  

/* API Design:
 *          This product was designed to contain and execute multiple tasks, 
 *          timed by user specifications. the product creates, schedule, and
 *          executes tasks. 
 *
 *  */                

#include <stddef.h> /* size_t */

#include "uid.h" /* UID_t */
#include "task.h" /* task_t */


typedef struct sched sched_t;



/* Description: Creates a scheduler container.
 *
 *        
 * Return Value: 
 *         pointer to the created scheduler.
*/
sched_t *SCHCreate(void);


/* Description: Destroys a scheduler container.
 *
 * Params: sched - pointer to the to be destroyed scheduler.
 *           
 * Return Value: none.
 *         
*/
void SCHDestroy(sched_t *sched);


/* Description: add a new task to the scheduler (by it's time).
 *
 * Params: sched - scheduler to insert.
 *         act_func  - user function of the task.
 *         param - optional parameter for the user function.
 *         interval - the interval between two executions of the functions.
 *        
 * Return Value: the UID of the added task.
 *         
*/
UID_t SCHAdd(sched_t *sched, task_function_t act_func, void *param,
                                                          size_t interval);

/* Description: removes a specific task from the schduler by UID.
 *
 * Params: ched - scheduler to remove from.
 *         task_id - task to remove.
 *        
 * Return Value: status - 0 - success, 1 - fail.
 *         
*/
int SCHRemove(sched_t *sched, UID_t task_id);


/* Description: activate the scheduler (execute tasks).
 *
 * Params: sched - the scheduler to activate.
 *    
 * Return Value: status - 0: no tasks left in sheduler.
 *                        1: terminated by flag.
 *                       -1: Error in setup of the run.
*/
int SCHRun(sched_t *sched);


/* Description: raises the stop flag of the scheduler.
 *
 * Params: sched - the scheduler to stop.
 *                 
*/
void SCHStop(sched_t *sched);


/* Description: counts number of tasks in scheduler.
 *
 * Params: sched - the scheduler.
 *        
 * Return Value: number of tasks in scheduler.
 *         
*/
size_t SCHSize(const sched_t *sched);


/* Description: checks whether the scheduler is empty or not.
 *
 * Params:sched - the scheduler.
 *        
 * Return Value: Boolean
 *         
*/
int SCHIsEmpty(const sched_t *sched);


#endif /* _SCHEDULER_H */
