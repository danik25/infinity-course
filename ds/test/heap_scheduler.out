/home/xyz/git/dani-benjamin/ds/heap_scheduler.out